### PR TITLE
cleanup!: require C++ >= 14

### PR DESCRIPTION
**Pull request by coryan**
Starting on the next release, we will require C++ >= 14. I also updated
the minimum GCC and Clang versions, to reflect what we actually support
and test with.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9339)

### CODE REVIEW

File: .bazelrc [hunk 1]
```diff
@@ -15,6 +15,9 @@
 # NOTE: This file is not loaded if google-cloud-cpp is built as part of a
 # larger project.
 
+build:linux --cxxopt=-std=c++14
+build:macos --cxxopt=-std=c++14
+
 # To workaround a bug in Bazel [1], gRPC requires this flag on macOS builds,
 # and there is (AFAIK) no way to "inherit" their definitions.
 #   [1]: https://github.com/bazelbuild/bazel/issues/4341
```
Comment (devbww) [inline .bazelrc:23]: Nit: Unnecessary difference in ordering between these two blocks here and in the generator/quickstart .bazelrc files?

Comment (coryan) [reply .bazelrc]: Fixed

File: google/cloud/accessapproval/quickstart/.bazelrc [hunk 2]
```diff
@@ -17,6 +17,19 @@
 #   [1]: https://github.com/bazelbuild/bazel/issues/4341
 build --copt=-DGRPC_BAZEL_BUILD
 
+# Use host-OS-specific config lines from bazelrc files.
+build --enable_platform_specific_config=true
+
+# The project requires C++ >= 14. By default Bazel adds `-std=c++0x` which
+# disables C++14 features, even if the compilers defaults to C++ >= 14
+# Use host-OS-specific config lines from bazelrc files.
+build --enable_platform_specific_config=true
```
Comment (devbww) [inline google/cloud/accessapproval/quickstart/.bazelrc:26]: Duplication.

Comment (coryan) [reply google/cloud/accessapproval/quickstart/.bazelrc]: Thanks.  I found that in only two files, and fixed both.

File: google/cloud/bigtable/mutation_batcher_test.cc [hunk 3]
```diff
@@ -138,7 +139,7 @@ class MutationBatcherTest : public bigtable::testing::TableTestFixture {
   MutationBatcherTest()
       : TableTestFixture(
             CompletionQueue(std::make_shared<FakeCompletionQueueImpl>())),
-        batcher_(new MutationBatcher(table_)) {}
+        batcher_(absl::make_unique<MutationBatcher>(table_)) {}
```
Comment (devjgm) [inline google/cloud/bigtable/mutation_batcher_test.cc:142]: Did you intend to use `std::make_unique` since C++14 is the min now?

**Review by devjgm — APPROVED** [review]
Exciting change. First major version bump. C++14. Cool :-)

File: google/cloud/bigtable/mutation_batcher_test.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/bigtable/mutation_batcher_test.cc:142]: I think we should, but I am punting on that discussion until later.  We decided not to go back and remove C++11-ism, at least not initially.

File: google/cloud/accessapproval/quickstart/.bazelrc [hunk 4]
```diff
@@ -17,6 +17,17 @@
 #   [1]: https://github.com/bazelbuild/bazel/issues/4341
 build --copt=-DGRPC_BAZEL_BUILD
 
+# Use host-OS-specific config lines from bazelrc files.
+build --enable_platform_specific_config=true
+
+# Use host-OS-specific config lines from bazelrc files.
+build --enable_platform_specific_config=true
```
Comment (devbww) [inline google/cloud/accessapproval/quickstart/.bazelrc:24]: Duplication.

File: google/cloud/storage/quickstart/.bazelrc [hunk 5]
```diff
@@ -17,6 +17,17 @@
 #   [1]: https://github.com/bazelbuild/bazel/issues/4341
 build --copt=-DGRPC_BAZEL_BUILD
 
+# Use host-OS-specific config lines from bazelrc files.
+build --enable_platform_specific_config=true
+
+# Use host-OS-specific config lines from bazelrc files.
+build --enable_platform_specific_config=true
```
Comment (devbww) [inline google/cloud/storage/quickstart/.bazelrc:24]: Duplication.

File: .bazelrc [hunk 6]
```diff
@@ -28,8 +28,9 @@ build --enable_platform_specific_config=true
 build:linux --cxxopt=-std=c++14
 build:macos --cxxopt=-std=c++14
 
-# Do not create the convenience links, they are inconvenient when the build runs
-# inside a docker image.
+# Do not create the convenience links, they are inconvenient when the build
```
Comment (devbww) [inline .bazelrc:31]: > I even fixed some comments for extra consistency.

Thanks.  I hesitate to suggest `s/links, they/links. They/`.  Your choice, depending upon how tired you are of non-functional changes in myriad files.

Comment (devbww) [reply .bazelrc]: LGTM.

**Review by devbww — APPROVED** [review]

**Pull request merged**